File: python/1b.py
#!/usr/bin/python

import itertools
import sys

INPUT = 'R4, R5, L5, L5, L3, R2, R1, R1, L5, R5, R2, L1, L3, L4, R3, L1, L1, R2, R3, R3, R1, L3, L5, R3, R1, L1, R1, R2, L1, L4, L5, R4, R2, L192, R5, L2, R53, R1, L5, R73, R5, L5, R186, L3, L2, R1, R3, L3, L3, R1, L4, L2, R3, L5, R4, R3, R1, L1, R5, R2, R1, R1, R1, R3, R2, L1, R5, R1, L5, R2, L2, L4, R3, L1, R4, L5, R4, R3, L5, L3, R4, R2, L5, L5, R2, R3, R5, R4, R2, R1, L1, L5, L2, L3, L4, L5, L4, L5, L1, R3, R4, R5, R3, L5, L4, L3, L1, L4, R2, R5, R5, R4, L2, L4, R3, R1, L2, R5, L5, R1, R1, L1, L5, L5, L2, L1, R5, R2, L4, L1, R4, R3, L3, R1, R5, L1, L4, R2, L3, R5, R3, R1, L3'


class Vector(object):
    def __init__(self, x, y):
        super(Vector, self).__setattr__('x', x)
        super(Vector, self).__setattr__('y', y)

    def __add__(self, vector):
        return Vector(self.x + vector.x, self.y + vector.y)

    def __mul__(self, value):
        return Vector(self.x * value, self.y * value)

    def __setattr__(self, k, v):
        raise RuntimeError('class Vector is immutable')

    def __hash__(self):
        return hash((self.x, self.y))

    def __eq__(self, vector):
        return self.x == vector.x and self.y == vector.y

    def manhattan(self):
        return abs(self.x) + abs(self.y)


def main():
    # Where we are in the city
    position = Vector(0, 0)
    # Start pointing north, python doesn't support previous iteration of cycles
    # so we maintain an index into the current direction we are facing
    directions = [Vector(0, 1), Vector(1, 0), Vector(0, -1), Vector(-1, 0)]
    index = 0
    # Process the input into a list of tuples of index increment and magnitude
    moves = [(1 if x[0] == 'R' else -1, int(x[1:])) for x in INPUT.split(', ')]
    # Record where we have visited
    visited = set([position])
    for direction, magnitude in moves:
        # Modulo 4 to keep the index in bounds
        index = (index + direction) % 4
        # Move in the desired direction the correct number of times
        for _ in range(0, magnitude):
            position += directions[index]
            if position in visited:
                # Answer is the manhattan distance from the start point
                print position.manhattan()
                sys.exit()
            visited.add(position)


if __name__ == '__main__':
    main()

# vi: ts=4 et:

File: c++/1b.cc
#include <stdlib.h>
#include <string>
#include <sstream>
#include <vector>
#include <iostream>
#include <set>
#include <boost/algorithm/string/classification.hpp>
#include <boost/algorithm/string/split.hpp>

using namespace std;

class CycleIterator {
public:
  CycleIterator(unsigned int size)
  : size(size), curr(0)
  {}

  void operator++(int) {
    curr++;
    curr %= size;
  }

  void operator--(int) {
    curr--;
    curr %= size;
  }

  unsigned int operator*() const {
    return curr;
  }

private:
  unsigned int size;
  unsigned int curr;
};

class Vector2i {
public:
  Vector2i(int x = 0, int y = 0)
  : x(x), y(y)
  {}

  inline Vector2i& operator+=(const Vector2i& v) {
    x += v.x;
    y += v.y;
    return *this;
  }

  inline Vector2i operator*(unsigned int i) const {
    return Vector2i(x * i, y * i);
  }

  inline unsigned int manhattan() const {
    return ::abs(x) + ::abs(y);
  };

  inline bool operator<(const Vector2i& v) const {
    if(x != v.x)
      return x < v.x;
    return y < v.y;
  }

  int x;
  int y;
};

int main() {
  string input("R4, R5, L5, L5, L3, R2, R1, R1, L5, R5, R2, L1, L3, L4, R3, L1, L1, R2, R3, R3, R1, L3, L5, R3, R1, L1, R1, R2, L1, L4, L5, R4, R2, L192, R5, L2, R53, R1, L5, R73, R5, L5, R186, L3, L2, R1, R3, L3, L3, R1, L4, L2, R3, L5, R4, R3, R1, L1, R5, R2, R1, R1, R1, R3, R2, L1, R5, R1, L5, R2, L2, L4, R3, L1, R4, L5, R4, R3, L5, L3, R4, R2, L5, L5, R2, R3, R5, R4, R2, R1, L1, L5, L2, L3, L4, L5, L4, L5, L1, R3, R4, R5, R3, L5, L4, L3, L1, L4, R2, R5, R5, R4, L2, L4, R3, R1, L2, R5, L5, R1, R1, L1, L5, L5, L2, L1, R5, R2, L4, L1, R4, R3, L3, R1, R5, L1, L4, R2, L3, R5, R3, R1, L3");

  vector<string> inputs;
  boost::split(inputs, input, boost::is_any_of(", "), boost::token_compress_on);

  Vector2i position;

  const Vector2i moves[4] = {Vector2i(0, 1), Vector2i(1, 0), Vector2i(0, -1), Vector2i(-1, 0)};
  CycleIterator i(4);

  set<Vector2i> visited;

  bool done = false;
  for(auto move = inputs.begin(); move != inputs.end() && !done; move++) {
    char direction;
    unsigned int magnitude;
    istringstream stream(*move);
    stream >> direction >> magnitude;

    if(direction == 'R')
      i++;
    else
      i--;

    for(unsigned int j = 0; j < magnitude; j++) {
      position += moves[*i];
      if(visited.find(position) != visited.end()) {
        done = true;
        break;
      }
      visited.insert(position);
    }
  }

  cout << position.manhattan() << endl;

  return 0;
}
